arch: arm: support built-in stack protection in user mode for ARMv8m

ioannisg opened the pull request:
This commit extends the support of built-in stack
protection (i.e. based on the stack pointer limit registers)
in ARMv8-M, so that it works with user mode. The commit
does the following:
- enables and disables the stack limit checking on the
privileged stack, inside thread system calls,
- does not enable stack limit checking in context switch,
if the thread is in user mode.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: support built-in stack protection in user mode for ARMv8m

This commit extends the support of built-in thread stack
protection (i.e. based on the stack pointer limit registers)
in ARMv8-M, so that it works with user mode. The commit
does the following:
- enables and disables the stack limit checking on the
privileged stack, inside thread system calls, and upon
entering user mode,
- does not enable stack limit checking in context switch,
if the thread is in user mode.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

pizi-nordic reviewed (changes requested):
There is some risk to crash the system (discussed with @ioannisg offline).

ioannisg committed:
arch: arm: typos and style fixes in userspace.S

Some minor typo fixes in inline comments within
arch/arm/core/userspace.S.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@pizi-nordic I addressed your concerns inthis PR. Update:
- PSPLIM is cleared before any updates to PSP (and then adjusted accordingly)
- configure_builtin_stack_guard() inline documentation is  enhanced
- PSPLIM is set to priv mode upon userspace enter

pizi-nordic commented on arch/arm/core/thread.c:
```diff
@@ -142,12 +142,21 @@ FUNC_NORETURN void _arch_user_mode_enter(k_thread_entry_t user_entry,
 #if defined(CONFIG_USERSPACE)
+	if (thread->arch.mode & CONTROL_nPRIV_Msk) {
+		/* Only configure stack limit for threads in privileged mode
+		 * (i.e supervisor threads or user threads doing system call).
```
IMHO this explains it better:
... (i.e supervisor threads or user threads currently executing system call) ...

ioannisg commented:
@andrewboie could you take a look?
